src/acomms/modemdriver/popoto_driver.cpp

```diff
@@ -132,133 +150,134 @@ void goby::acomms::popotoDriver::handle_initiate_transmission(
 
     std::stringstream message;
-    message << "playstart "
-                <<  msg.GetExtension(popoto::protobuf::transmission).file_location() << " "
-                << msg.GetExtension(popoto::protobuf::transmission).transmit_power() << "\n";
+    message << "playstart " << msg.GetExtension(popoto::protobuf::transmission).file_location()
+            << " " << msg.GetExtension(popoto::protobuf::transmission).transmit_power() << "\n";
 
-    // send over the wire 
+    // send over the wire
     signal_and_write("playstop\n"); // need to make sure nothing else is playing
     signal_and_write(message.str());
-
 }
 //--------------------------------------- send_ping ------------------------------------------------------------
 // Send a ping
-void goby::acomms::popotoDriver::send_ping(protobuf::ModemTransmission& msg)
+void goby::acomms::PopotoDriver::send_ping(protobuf::ModemTransmission& msg)
 {
-    std::cerr << msg.DebugString() << std::endl; 
+    glog.is(DEBUG1) && glog << msg.DebugString() << std::endl;
 
     std::stringstream message;
     message << "ping " << msg.GetExtension(popoto::protobuf::transmission).transmit_power() << "\n";
-    
+
     signal_and_write(message.str());
-          
 }
 
 // ------------------------------------- Send -------------------------------------------------
-void goby::acomms::popotoDriver::send(protobuf::ModemTransmission& msg)
+void goby::acomms::PopotoDriver::send(protobuf::ModemTransmission& msg)
 {
-    int dest = msg.dest();
-    
-    std::stringstream raw1; 
+    // Set bitrate
+    // Bitrates with Popoto modem: map these onto 0-5
+    std::vector<std::string> rate_to_speed{"setRate80\n",   "setRate640\n",  "setRate1280\n",
+                                           "setRate2560\n", "setRate5120\n", "setRate10240\n"};
+    int min_rate = 0, max_rate = rate_to_speed.size() - 1;
+    int rate = msg.rate();
+
+    if (rate < min_rate || rate > max_rate)
+    {
+        glog.is(WARN) && glog << "Invalid rate, must be between " << min_rate << " and " << max_rate
+                              << ". Using rate: " << min_rate << std::endl;
+        rate = min_rate;
+    }
+    signal_and_write(rate_to_speed[rate]);
+
+    int dest = (msg.dest() == goby::acomms::BROADCAST_ID) ? POPOTO_BROADCAST_ID : msg.dest();
+    std::stringstream raw1;
     raw1 << "setvaluei RemoteID " << dest << "\n";
     signal_and_write(raw1.str());
-    
-    std::stringstream out_msg;
-    int d = msg.frame_start(); 
-    out_msg << "ACK: " << d;
-    protobuf::ModemRaw raw_out;
-    raw_out.set_raw(out_msg.str());
-    ModemDriverBase::signal_raw_incoming(raw_out);
-
-    // Pass the binary packet to comma to convert it to csv
-    char buf[1024];
-    std::string ss;
-    msg.clear_slot_seconds();
-    msg.clear_time();    
-    msg.clear_slot_index();
-
-    if( msg.type() != protobuf::ModemTransmission::ACK)
+
+    auto goby_header = CreateGobyHeader(msg);
```

Comment:
Protobuf encoding is very inefficient for acoustic modems, which is why we don't use it on the wire (e.g. SerializeToString). Here we can get most of what we need from the Popoto header, but we have a few extra bits of information (DATA/ACK, ack_requested) that we need to have. I've implemented this as a single prefix byte returned by CreateGobyHeader.